### PR TITLE
YR/-BC-104-New-format-validation/CIAC-10001

### PR DESCRIPTION
**Pull request by RosenbergYehuda**
## Related Issues
fixes: https://jira-dc.paloaltonetworks.com/browse/CIAC-10001
fixes: https://jira-dc.paloaltonetworks.com/browse/CIAC-10114
fixes: https://jira-dc.paloaltonetworks.com/browse/CIAC-9694

## Description

Convert half of BC 104 to the new format. This validation ensures that the names of existing commands or arguments have not been changed.
Furthermore, a new validation, BC110, was introduced following the split of BC104. This validation ensures that non-required arguments are not converted into required ones and that new arguments cannot be made mandatory.

### CODE REVIEW

File: demisto_sdk/commands/validate/validators/BC_validators/BC104_have_commands_or_args_name_changed.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+class HaveCommandsOrArgsNameChangedValidator(BaseValidator[ContentTypes]):
+    error_code = "BC104"
+    description = "Check if the command name or argument name has been changed."
+    rationale = "If an existing command or argument has been renamed, it will break backward compatibility"
+    error_message = "Possible backward compatibility break: Your updates to this file contain changes {unique_message} Please undo the changes."
+    related_field = "name"  # TODO - what is the field name?
+    is_auto_fixable = False
+    expected_git_statuses = [GitStatuses.MODIFIED]
+    related_file_type = [RelatedFileType.YML]
```
Comment (YuvHayun) [inline demisto_sdk/commands/validate/validators/BC_validators/BC104_have_commands_or_args_name_changed.py:25]:
```suggestion
    related_field = "script.commands.arguments.name"
    expected_git_statuses = [GitStatuses.MODIFIED]
```

File: demisto_sdk/commands/validate/validators/BC_validators/BC104_have_commands_or_args_name_changed.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+            if args_diff_per_command_summary:
+                results.append(
+                    ValidationResult(
+                        validator=self,
+                        message=self.error_message.format(
+                            unique_message=f"to the names of existing arguments: {' '.join(args_diff_per_command_summary)}",
+                        ),
+                        content_object=content_item,
+                    )
+                )
```
Comment (YuvHayun) [inline demisto_sdk/commands/validate/validators/BC_validators/BC104_have_commands_or_args_name_changed.py:83]: I don't like the idea of returning 2 messages from the same validation.
Let's either find a way to return one validationResult, or split this validation into two validations.

File: demisto_sdk/commands/validate/tests/BC_validators_test.py [hunk 3]
```diff
@@ -891,3 +897,138 @@ def test_IsContextPathChangedValidator_remove_command():
         f"Command {command_name} has been removed from the integration. This is a breaking change, and is not allowed."
         in errors[0].message
     )
+
+
+def setup_integration_with_3_commands_and_5_args():
```
Comment (YuvHayun) [inline demisto_sdk/commands/validate/tests/BC_validators_test.py:902]: Let's remove this function, I don't think that a helper function to init the args is a good idea.
Also, it's done wrong here, there's a better way to init an integration with 3 commands.
see 'test_IsContainingMultipleDefaultArgsValidator_is_valid' in IN_Validators as a reference.

File: demisto_sdk/commands/validate/tests/BC_validators_test.py [hunk 4]
```diff
@@ -891,3 +897,138 @@ def test_IsContextPathChangedValidator_remove_command():
...
+    new_content_item = setup_integration_with_3_commands_and_5_args()
+    new_content_item.commands[1].name = "new-command_1"
+    new_content_item.commands[1].args[0].name = "new_arg_1"
+    new_content_item.commands[2].args[1].name = "new_arg_2"
+
+    # Setup old content item with original command and argument names
+    old_content_item = setup_integration_with_3_commands_and_5_args()
+    old_content_item.commands[1].name = "old-command_1"
+    old_content_item.commands[1].args[0].name = "old_arg_1"
+    old_content_item.commands[2].args[1].name = "old_arg_2"
```
Comment (YuvHayun) [inline demisto_sdk/commands/validate/tests/BC_validators_test.py:939]: It's done wrong here, there's a better way to init an integration with 3 commands.
see 'test_IsContainingMultipleDefaultArgsValidator_is_valid' in IN_Validators as a reference.
You can init the old content object and then use copy.deepcopy to create the new object and alter it.

File: demisto_sdk/commands/validate/tests/BC_validators_test.py [hunk 5]
```diff
@@ -891,3 +897,138 @@ def test_IsContextPathChangedValidator_remove_command():
...
+
+    new_content_item.commands.append(create_integration_object().commands[0])
+    new_content_item.commands[2].args.append(
+        create_integration_object().commands[0].args[0]
+    )
+    new_content_item.commands[3].name = "new_command_3"
+    new_content_item.commands[2].args[2].name = "new_arg_2"
+
+    # Setup old content item with original command and argument names
+    old_content_item = setup_integration_with_3_commands_and_5_args()
```
Comment (YuvHayun) [inline demisto_sdk/commands/validate/tests/BC_validators_test.py:977]: It's done wrong here, there's a better way to init an integration with 3 commands.
see 'test_IsContainingMultipleDefaultArgsValidator_is_valid' in IN_Validators as a reference.
You can init the old content object and then use copy.deepcopy to create the new object and alter it.

File: demisto_sdk/commands/validate/tests/BC_validators_test.py [hunk 6]
```diff
@@ -891,3 +897,138 @@ def test_IsContextPathChangedValidator_remove_command():
...
+    """
+    new_content_item = setup_integration_with_3_commands_and_5_args()
+    new_content_item.commands[2].args.append(
+        create_integration_object().commands[0].args[0]
+    )
+    # change existing arg to be required
+    new_content_item.commands[1].args[0].required = True
+    # add new required arg
+    new_content_item.commands[2].args[2].required = True
+    old_content_item = setup_integration_with_3_commands_and_5_args()
```
Comment (YuvHayun) [inline demisto_sdk/commands/validate/tests/BC_validators_test.py:1001]: It's done wrong here, there's a better way to init an integration with 3 commands.
see 'test_IsContainingMultipleDefaultArgsValidator_is_valid' in IN_Validators as a reference.
You can init the old content object and then use copy.deepcopy to create the new object and alter it.

File: demisto_sdk/commands/validate/tests/BC_validators_test.py [hunk 7]
```diff
@@ -891,3 +897,138 @@ def test_IsContextPathChangedValidator_remove_command():
...
+    new_content_item = setup_integration_with_3_commands_and_5_args()
+    # add a new command
+    new_content_item.commands.append(create_integration_object().commands[0])
+    # change existing arg to be required but with a default value
+    new_content_item.commands[1].args[0].required = True
+    new_content_item.commands[1].args[0].defaultvalue = "test"
+    # change arg in the new command to be required
+    new_content_item.commands[3].args[0].required = True
+
+    old_content_item = setup_integration_with_3_commands_and_5_args()
```
Comment (YuvHayun) [inline demisto_sdk/commands/validate/tests/BC_validators_test.py:1031]: It's done wrong here, there's a better way to init an integration with 3 commands.
see 'test_IsContainingMultipleDefaultArgsValidator_is_valid' in IN_Validators as a reference.
You can init the old content object and then use copy.deepcopy to create the new object and alter it.

File: demisto_sdk/commands/validate/tests/BC_validators_test.py [hunk 8]
```diff
@@ -891,3 +897,138 @@ def test_IsContextPathChangedValidator_remove_command():
...
+    # change existing arg to be required
+    new_content_item.commands[1].args[0].required = True
+    # add new required arg
+    new_content_item.commands[2].args[2].required = True
+    old_content_item = setup_integration_with_3_commands_and_5_args()
+
+    create_old_file_pointers([new_content_item], [old_content_item])
+    res = NewRequiredArgumentValidator().is_valid([new_content_item])
+    assert (
+        "in command 'test-command_1' you have added a new required argument: 'from_date'.in command 'test-command_2' you have added a new required argument: 'from_date'. Please undo the changes."
```
Comment (YuvHayun) [inline demisto_sdk/commands/validate/tests/BC_validators_test.py:1006]:
```suggestion
        "in command 'test-command_1' you have added a new required argument: 'from_date'.in command 'test-command_2' you have added a new required argument: 'from_date'. Please undo the changes."
```
''from_date'.in command '
Make sure to alter the validation code to create space between the '.' and 'in'
I think splitting by \n would be more readable in that case.

File: demisto_sdk/commands/validate/tests/BC_validators_test.py [hunk 9]
```diff
@@ -891,3 +897,138 @@ def test_IsContextPathChangedValidator_remove_command():
...
+    results = HaveCommandsOrArgsNameChangedValidator().is_valid([new_content_item])
+
+    # Assert error messages
+    assert (
+        "contain changes to the names of the following existing commands: old-command_1. Please undo the changes."
+        in results[0].message
+    )
+    assert (
+        "contain changes to the names of existing arguments: In command 'test-command_2' the following arguments have been changed: old_arg_2. Please undo the changes."
+        in results[1].message
```
Comment (YuvHayun) [inline demisto_sdk/commands/validate/tests/BC_validators_test.py:952]: As mentioned in the validation itself, I don't like the idea of splitting into two different validationResults here.

**Review by YuvHayun — CHANGES_REQUESTED** [review]
Good job!
See my comments.

File: demisto_sdk/commands/validate/validators/BC_validators/BC110_new_required_argument.py [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+
+ContentTypes = Integration
+
+
+class NewRequiredArgumentValidator(BaseValidator[ContentTypes]):
+    error_code = "BC110"
+    description = (
+        "Validate that no new *required* argument are added to an existing command."
+    )
+    rationale = "Adding a new argument to an existing command and defining it as *required* or changing an non required argument to be required will break backward compatibility."
```
[the remark this answers is not in the document]
Comment (RosenbergYehuda) [reply demisto_sdk/commands/validate/validators/BC_validators/BC110_new_required_argument.py:22]: Done

File: demisto_sdk/commands/validate/validators/BC_validators/BC110_new_required_argument.py [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+    error_message = "Possible backward compatibility break: You have added the following new *required* arguments: {custom_message} Please undo the changes."
+    related_field = ""  # TODO - what is the field name?
```
[the remark this answers is not in the document]
Comment (RosenbergYehuda) [reply demisto_sdk/commands/validate/validators/BC_validators/BC110_new_required_argument.py:24]: Done

File: demisto_sdk/commands/validate/validators/BC_validators/BC104_have_commands_or_args_name_changed.py (hunk 2 above)
Comment (RosenbergYehuda) [reply demisto_sdk/commands/validate/validators/BC_validators/BC104_have_commands_or_args_name_changed.py:83]: fixed

File: demisto_sdk/commands/validate/tests/BC_validators_test.py (hunk 9 above)
Comment (RosenbergYehuda) [reply demisto_sdk/commands/validate/tests/BC_validators_test.py:952]: done

File: demisto_sdk/commands/validate/tests/BC_validators_test.py (hunk 4 above)
Comment (RosenbergYehuda) [reply demisto_sdk/commands/validate/tests/BC_validators_test.py:939]: done

File: demisto_sdk/commands/validate/tests/BC_validators_test.py (hunk 5 above)
Comment (RosenbergYehuda) [reply demisto_sdk/commands/validate/tests/BC_validators_test.py:977]: done

File: demisto_sdk/commands/validate/tests/BC_validators_test.py (hunk 6 above)
Comment (RosenbergYehuda) [reply demisto_sdk/commands/validate/tests/BC_validators_test.py:1001]: done

File: demisto_sdk/commands/validate/tests/BC_validators_test.py (hunk 7 above)
Comment (RosenbergYehuda) [reply demisto_sdk/commands/validate/tests/BC_validators_test.py:1031]: done

File: demisto_sdk/commands/validate/tests/BC_validators_test.py (hunk 3 above)
Comment (RosenbergYehuda) [reply demisto_sdk/commands/validate/tests/BC_validators_test.py:902]: You rather it as a global? and not a function?

File: demisto_sdk/commands/validate/tests/BC_validators_test.py (hunk 8 above)
Comment (RosenbergYehuda) [reply demisto_sdk/commands/validate/tests/BC_validators_test.py:1006]: done

File: demisto_sdk/commands/validate/sdk_validation_config.toml [hunk 12]
```diff
@@ -25,7 +25,7 @@ select = [
     "BA100", "BA101", "BA105", "BA106", "BA110", "BA111", "BA114", "BA116", "BA118", "BA119", "BA124", "BA126",
     "PA100", "PA101", "PA102", "PA103", "PA104", "PA105", "PA107", "PA108", "PA109", "PA111", "PA113", "PA115", "PA117", "PA118", "PA119", "PA120",
     "PA121", "PA123", "PA125", "PA127", "PA128", "PA130",
-    "BC100", "BC101", "BC102", "BC105", "BC108",
+    "BC100", "BC101", "BC102", "BC104", "BC105", "BC108", "BC110"
```
Comment (YuvHayun) [inline demisto_sdk/commands/validate/sdk_validation_config.toml:28]:
```suggestion
    "BC100", "BC101", "BC102", "BC104", "BC105", "BC108", "BC110",
```

File: demisto_sdk/commands/validate/validators/BC_validators/BC110_new_required_argument.py [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+
+class NewRequiredArgumentValidator(BaseValidator[ContentTypes]):
+    error_code = "BC110"
+    description = (
+        "Validate that no new *required* argument are added to an existing command."
+    )
+    rationale = "Adding a new argument to an existing command and defining it as *required* or changing an non-required argument to be required will break backward compatibility."
+    error_message = "Possible backward compatibility break: You have added the following new *required* arguments: {custom_message} Please undo the changes."
+    related_field = "script.commands.arguments"
+    is_auto_fixable = False
```
Comment (YuvHayun) [inline demisto_sdk/commands/validate/validators/BC_validators/BC110_new_required_argument.py:24]:
```suggestion
```

File: demisto_sdk/commands/validate/validators/BC_validators/BC104_have_commands_or_args_name_changed.py [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+ContentTypes = Integration
+
+
+class HaveCommandsOrArgsNameChangedValidator(BaseValidator[ContentTypes]):
+    error_code = "BC104"
+    description = "Check if the command name or argument name has been changed."
+    rationale = "If an existing command or argument has been renamed, it will break backward compatibility"
+    error_message = "Possible backward compatibility break: Your updates to this file contain changes {final_message} Please undo the changes."
+    related_field = "script.commands.arguments.name"
+    is_auto_fixable = False
```
Comment (YuvHayun) [inline demisto_sdk/commands/validate/validators/BC_validators/BC104_have_commands_or_args_name_changed.py:22]:
```suggestion
```

File: demisto_sdk/commands/validate/tests/BC_validators_test.py [hunk 15]
```diff
@@ -43,6 +50,53 @@
...
+                        "description": "nothing description.",
+                    },
+                ],
+                "outputs": [],
+            },
+        ]
+    ],
+)
+
+
```
Comment (YuvHayun) [inline demisto_sdk/commands/validate/tests/BC_validators_test.py:99]: Please move it to the parametrize.
It's a bad practice to init it as a constant from outside the tests.

**Review by YuvHayun — APPROVED** [review]
Gj, minor comments.
@JasBeilin approved from my side.

File: demisto_sdk/commands/validate/tests/tools_test.py [hunk 16]
```diff
@@ -87,3 +88,20 @@ def test_collect_all_inputs_from_inputs_section(content_item, expected_result):
...
+
+
+def test_compare_lists():
+    """
+    Given:
+        - Two lists with some elements in common and some that are not.
+    When:
+        - Running compare_lists
+    Then:
+        - Return a list of the elements that are in the sub_list but not in the main_list, including duplicates if not all elements are in the main_list.
```
Comment (JasBeilin) [inline demisto_sdk/commands/validate/tests/tools_test.py:100]: lets improve this line.

File: demisto_sdk/commands/validate/validators/BC_validators/BC104_have_commands_or_args_name_changed.py [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+    ValidationResult,
+)
+
+ContentTypes = Integration
+
+
+class HaveCommandsOrArgsNameChangedValidator(BaseValidator[ContentTypes]):
+    error_code = "BC104"
+    description = "Check if the command name or argument name has been changed."
+    rationale = "If an existing command or argument has been renamed, it will break backward compatibility"
```
Comment (JasBeilin) [inline demisto_sdk/commands/validate/validators/BC_validators/BC104_have_commands_or_args_name_changed.py:19]: PLease add the diff from BC111&BC103

**Review by JasBeilin — CHANGES_REQUESTED** [review]
great, think about making it simpler.

File: demisto_sdk/commands/validate/tests/BC_validators_test.py [hunk 18]
```diff
@@ -54,7 +54,8 @@
 
 
 # Create a new content item with 3 commands with unique names. all commands have only 1 argument except the third command which has 2 arguments.
-generic_integration_with_3_commands_and_4_args = create_integration_object(
+
+GENERIC_INTEGRATION_WITH_3_COMMANDS_AND_4_ARGS = create_integration_object(
```
Comment (JasBeilin) [inline demisto_sdk/commands/validate/tests/BC_validators_test.py:58]: Why is the number of args & command important here? cant we call it just DUMMY_INTEGRATION?

File: demisto_sdk/commands/validate/tools.py [hunk 19]
```diff
@@ -153,6 +153,10 @@ def find_command(commands: List[Command], command_to_find: str) -> Optional[Comm
 def compare_lists(sub_list: List[str], main_list: List[str]) -> List[str]:
     """
     Compares two lists and returns the elements that are in the sub list but not in the main list, including duplicates.
+    for example:
+    sub_list = ['a', 'b', 'c', 'c', 'd']
+    main_list = ['a', 'b', 'c']
+    compare_lists(sub_list, main_list) -> ['c', 'd']
```
Comment (JasBeilin) [inline demisto_sdk/commands/validate/tools.py:159]:
```suggestion
    compare_lists(['a', 'b', 'c', 'c', 'd'], ['a', 'b', 'c', 'c', 'd']) -> ['c', 'd']
```

File: demisto_sdk/commands/validate/tests/tools_test.py [hunk 20]
```diff
@@ -93,11 +93,14 @@ def test_collect_all_inputs_from_inputs_section(content_item, expected_result):
...
+        - A `main_list` containing the elements "a", "b", and "c".
+        - A `sub_list` containing the elements "a", "b", "b", and "d".
+
     When:
-        - Running compare_lists
+        - The function `compare_lists` is called with `sub_list` and `main_list` as arguments.
+
     Then:
-        - Return a list of the elements that are in the sub_list but not in the main_list, including duplicates if not all elements are in the main_list.
+        - The function should return a list containing the elements "b" and "d", which are present in `sub_list` but not in `main_list`.
```
Comment (JasBeilin) [inline demisto_sdk/commands/validate/tests/tools_test.py:103]:
```suggestion
        - Ensuring the value returned is a list containing the elements "b" and "d", which are present in `sub_list` but not in `main_list`.
```

**Review by JasBeilin — APPROVED** [review]
Great, See my comments.

File: demisto_sdk/commands/validate/tests/BC_validators_test.py (hunk 18 above)
Comment (RosenbergYehuda) [reply demisto_sdk/commands/validate/tests/BC_validators_test.py:58]: I think it helps to understand the test, especially when i an adding a new arg, so it is clear to know what did we have before and what we have now.

**Pull request merged**